### ISSUE TITLE
New rule: enforce-case-break

### ISSUE THREAD

**Issue by futagoza** [eslint/eslint#11179]
**Please describe what the rule should do:**

Should warn (or throw an error) when it encounters case blocks that don't contain a break statement.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```json
{
  "rules": {
    "enforce-case-break": [ "warn", {
        "shouldBeTheLastStatement": false,
        "allowOnlyABreak": true,
        "ignoreMultiCases": true,
        "ignoreDefaultCase": true,
        "ignoreLastCase": true,
    } ],
  }
}
```

With the above config (in recomended maybe? 👼), the following snippet should throw 1 warning:

```js
switch ( current_argv ) {

  case "--help":
    console.log( "no ;P" ); // <--- should warn about expecting a break statement

  case "-d": break; // <----------- ignored because allowOnlyABreak is true

  default:
    console.log( "hi" ); // <------ ignored because ignoreDefaultCase is true

  case "-v": // <------------------ ignored because ignoreMultiCases is true
  case "--version":
    console.log( 5 ); // <--------- ignored because ignoreLastCase is true

}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

I've been programming in JavaScript for 8 years now, and even though I knew that empty case statements fall to the next case block (usually called multi-cases it seems, and I have used these quite often), I just discovered that if a non-empty case block doesn't contain a break statement, it should also fall to the next case block. I don't know about you guys, but I'm sure I'm not alone on this train of thought 😆

Having this in core (and the above config in recomended) helps to ensure the desired behaviour of a switch block (in other words, ensure that someone doesn't add a case block that falls to the following case block when it shouldn't).

**Are you willing to submit a pull request to implement this rule?**

Forget about creating a new rule, I haven't even hacked on ESLint before. Either over the coming weekend or next week, I'll see if I can do this. If I can, and no one has picked it up, I'll give it a go.

**Comment by futagoza** [comment]
#### `allowOnlyABreak` option

This is because I've come across situations where I have a switch block that contains a default, but I want to ignore some cases while handling others.

#### `ignoreMultiCases` option

I was thinking, would it be better to not add this option and always ignore them?

#### `shouldBeTheLastStatement` option

If `true`, this would enforce the placing of a break statement to be at the end of a case block. By default though, I think this should be `false` so that we can _choose_ to fall to the next case block if desired.

**Comment by g-plane** [comment]
There are some existing rules doing this:
- `default-case`: https://eslint.org/docs/rules/default-case#require-default-case-in-switch-statements-default-case
- `no-fallthrough`: https://eslint.org/docs/rules/no-fallthrough#disallow-case-statement-fallthrough-no-fallthrough

**Comment by futagoza** [comment]
@g-plane it's funny how I actually came across `no-fallthrough` about a month ago but didn't register it in my mind because I was doing something else 😆

**Comment by platinumazure** [comment]
@ljharb Which internal rule are you referring to?

**Comment by ljharb** [comment]
@platinumazure oops, i think i confused this issue with another one. deleting my comments :-)
